File: capstone-full-/inventory.js
// Inventory Page JavaScript - PRODUCTION READY

// ===== MODAL ELEMENTS =====
const restockModal = document.getElementById('restockModal');
const editModal = document.getElementById('editModal');
const addIngredientModal = document.getElementById('addIngredientModal');

// ===== MODAL CONTROL FUNCTIONS =====
function openRestockModal() {
  loadIngredientsDropdown('ingredientSelect');
  restockModal.showModal();
}

function closeRestockModal() {
  restockModal.close();
  document.getElementById('restockForm').reset();
}

function openEditModal() {
  loadIngredientsDropdown('editIngredient');
  editModal.showModal();
}

function closeEditModal() {
  editModal.close();
  document.getElementById('editForm').reset();
}

function openAddIngredientModal() {
  addIngredientModal.showModal();
}

function closeAddIngredientModal() {
  addIngredientModal.close();
  document.getElementById('addIngredientForm').reset();
}

// ===== LOAD INGREDIENTS DROPDOWN =====
async function loadIngredientsDropdown(selectId) {
  const select = document.getElementById(selectId);
  select.innerHTML = '<option value="">Loading...</option>';
  
  try {
    const res = await fetch('http://127.0.0.1:5000/api/inventory');
    if (!res.ok) throw new Error('Failed to load ingredients');
    const data = await res.json();
    
    select.innerHTML = '<option value="">Select ingredient...</option>';
    data.forEach(item => {
      const option = document.createElement('option');
      option.value = item.item_id;
      option.textContent = item.item_name;
      option.dataset.stockLevel = item.stock_level;
      option.dataset.capacity = item.capacity;
      option.dataset.category = item.category || '';
      option.dataset.status = item.status || 'active';
      select.appendChild(option);
    });
  } catch (err) {
    console.error('Failed to load ingredients:', err);
    select.innerHTML = '<option value="">Failed to load</option>';
  }
}

// ===== HANDLE INGREDIENT SELECTION =====
document.getElementById('editIngredient').addEventListener('change', function() {
  const selected = this.options[this.selectedIndex];
  if (selected.value) {
    document.getElementById('currentStock').value = selected.dataset.stockLevel || 0;
    document.getElementById('requiredStock').value = selected.dataset.capacity || 0;
    document.getElementById('category').value = selected.dataset.category || '';
    
    const status = selected.dataset.status || 'active';
    document.getElementById('disableBtn').style.display = status === 'active' ? 'block' : 'none';
    document.getElementById('enableBtn').style.display = status === 'disabled' ? 'block' : 'none';
  }
});

// ===== INGREDIENT MANAGEMENT FUNCTIONS =====
async function disableIngredient() {
  const select = document.getElementById('editIngredient');
  const ingredientId = select.value;
  const ingredientName = select.options[select.selectedIndex].text;
  
  if (!ingredientId) return alert('Please select an ingredient first');
  if (!confirm(`Disable "${ingredientName}"? It will be hidden from active inventory.`)) return;

  try {
    const res = await fetch(`http://127.0.0.1:5000/api/inventory/${ingredientId}/disable`, { 
      method: 'POST' 
    });
    if (!res.ok) throw new Error('Failed to disable');
    
    alert(`"${ingredientName}" has been disabled! ✅`);
    closeEditModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to disable ingredient:', err);
    alert('Failed to disable ingredient. Please try again.');
  }
}

async function enableIngredient() {
  const select = document.getElementById('editIngredient');
  const ingredientId = select.value;
  const ingredientName = select.options[select.selectedIndex].text;
  
  if (!ingredientId) return alert('Please select an ingredient first');

  try {
    const res = await fetch(`http://127.0.0.1:5000/api/inventory/${ingredientId}/enable`, { 
      method: 'POST' 
    });
    if (!res.ok) throw new Error('Failed to enable');
    
    alert(`"${ingredientName}" has been re-enabled! ✅`);
    closeEditModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to enable ingredient:', err);
    alert('Failed to enable ingredient. Please try again.');
  }
}

async function deleteIngredient() {
  const select = document.getElementById('editIngredient');
  const ingredientId = select.value;
  const ingredientName = select.options[select.selectedIndex].text;
  
  if (!ingredientId) return alert('Please select an ingredient first');
  if (!confirm(`Permanently DELETE "${ingredientName}"? This cannot be undone!`)) return;

  try {
    const res = await fetch(`http://127.0.0.1:5000/api/inventory/${ingredientId}`, { 
      method: 'DELETE' 
    });
    if (!res.ok) throw new Error('Failed to delete');
    
    alert(`"${ingredientName}" has been deleted! ✅`);
    closeEditModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to delete ingredient:', err);
    alert('Failed to delete ingredient. Please try again.');
  }
}

// ===== FORM SUBMISSIONS =====

// Restock Form
document.getElementById('restockForm').addEventListener('submit', async (e) => {
  e.preventDefault();
  
  const ingredientId = document.getElementById('ingredientSelect').value;
  const quantity = parseFloat(document.getElementById('quantity').value);
  const unit = document.getElementById('unit').value;

  try {
    const res = await fetch('http://127.0.0.1:5000/api/inventory/restock', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ item_id: ingredientId, quantity, unit })
    });
    if (!res.ok) throw new Error('Failed to restock');
    
    alert('Restock logged successfully! ✅');
    closeRestockModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to log restock:', err);
    alert('Failed to log restock. Please try again.');
  }
});

// Edit Form
document.getElementById('editForm').addEventListener('submit', async (e) => {
  e.preventDefault();
  
  const ingredientId = document.getElementById('editIngredient').value;
  const stock_level = parseFloat(document.getElementById('currentStock').value);
  const capacity = parseFloat(document.getElementById('requiredStock').value);
  const category = document.getElementById('category').value;

  try {
    const res = await fetch(`http://127.0.0.1:5000/api/inventory/${ingredientId}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ stock_level, capacity, category })
    });
    if (!res.ok) throw new Error('Failed to update');
    
    alert('Ingredient updated successfully! ✅');
    closeEditModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to update ingredient:', err);
    alert('Failed to update ingredient. Please try again.');
  }
});

// Add Ingredient Form
document.getElementById('addIngredientForm').addEventListener('submit', async (e) => {
  e.preventDefault();
  
  const item_name = document.getElementById('newIngredientName').value;
  const stock_level = parseFloat(document.getElementById('newCurrentStock').value);
  const capacity = parseFloat(document.getElementById('newRequiredStock').value);
  const category = document.getElementById('newCategory').value;

  try {
    const res = await fetch('http://127.0.0.1:5000/api/inventory', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ item_name, stock_level, capacity, category })
    });
    if (!res.ok) throw new Error('Failed to add ingredient');
    
    alert(`"${item_name}" added successfully! ✅`);
    closeAddIngredientModal();
    setTimeout(loadInventory, 300);
  } catch (err) {
    console.error('Failed to add ingredient:', err);
    alert('Failed to add ingredient. Please try again.');
  }
});

// ===== MODAL BACKDROP CLOSE =====
[restockModal, editModal, addIngredientModal].forEach(modal => {
  modal.addEventListener('click', (e) => { 
    if (e.target === modal) modal.close(); 
  });
});

// ===== ESC KEY TO CLOSE =====
document.addEventListener('keydown', (e) => {
  if (e.key === 'Escape') document.querySelector('dialog[open]')?.close();
});

// ===== PROGRESS BAR HELPER =====
function getProgressClass(percent) {
  if (percent === 0) return 'progress-empty';
  if (percent <= 20) return 'progress-bad';
  if (percent <= 60) return 'progress-warn';
  return 'progress-good';
}

// ===== FALLBACK DATA (for offline mode) =====
const fallbackData = {
  perishable: [
    { name: "Milk", percent: 70 },
    { name: "Lettuce", percent: 10 },
    { name: "Chicken", percent: 100 },
    { name: "Tomato", percent: 40 },
    { name: "Egg", percent: 90 },
    { name: "Fresh Fruits", percent: 50 }
  ],
  semiperishable: [
    { name: "Coffee Beans", percent: 60 },
    { name: "Bread", percent: 20 },
    { name: "Coconut Milk", percent: 80 },
    { name: "Cocoa Powder", percent: 45 },
    { name: "Matcha Powder", percent: 30 },
    { name: "Mayonnaise", percent: 50 },
    { name: "Jam", percent: 70 },
    { name: "Cream", percent: 60 }
  ]
};

// ===== RENDER INVENTORY =====
function renderGroups(perishableArr, semiArr) {
  const perishableList = document.getElementById('perishable-list');
  const semiList = document.getElementById('semiperishable-list');
  if (!perishableList || !semiList) return;

  const makeRows = arr => arr.map(item => `
    <div class="ingredient-row">
      <span class="ingredient-name">${item.name}</span>
      <div class="progress-bar">
        <div class="progress-fill ${getProgressClass(item.percent)}" style="width:${item.percent}%"></div>
      </div>
    </div>
  `).join('');

  perishableList.innerHTML = perishableArr.length ? makeRows(perishableArr) : "<div class='muted'>No perishable items.</div>";
  semiList.innerHTML = semiArr.length ? makeRows(semiArr) : "<div class='muted'>No semi-perishable items.</div>";

  // Animate progress bars
  requestAnimationFrame(() => {
    document.querySelectorAll('.progress-fill').forEach((fill, i) => {
      const target = fill.style.width;
      fill.style.width = '0%';
      setTimeout(() => { fill.style.width = target; }, 80 + i * 20);
    });
  });
}

// ===== LOAD INVENTORY =====
async function loadInventory() {
  const perishableList = document.getElementById('perishable-list');
  const semiList = document.getElementById('semiperishable-list');
  if (!perishableList || !semiList) return;

  perishableList.innerHTML = "<div class='ingredient-row loading'><span class='ingredient-name muted'>Loading...</span></div>";
  semiList.innerHTML = "<div class='ingredient-row loading'><span class='ingredient-name muted'>Loading...</span></div>";

  try {
    const res = await fetch('http://127.0.0.1:5000/api/inventory');
    if (!res.ok) throw new Error('Bad response');
    const rows = await res.json();

    // Fallback category mapping (if database category is missing)
    const perishableNames = new Set([
      'milk', 'lettuce', 'chicken', 'tomato', 'bread', 'egg', 
      'lemon', 'mushroom', 'fresh fruits', 'cream', 'jam'
    ]);

    const perishable = [];
    const semi = [];

    rows.forEach(r => {
      if (r.status === 'disabled') return; // Skip disabled
      
      const rawName = (r.item_name || '').trim();
      const key = rawName.toLowerCase();
      const pct = r.capacity ? Math.min(100, Math.round((r.stock_level / r.capacity) * 100)) : 0;
      
      // Smart categorization: use database category, fallback to name-based
      const target = r.category
        ? (r.category.toLowerCase().includes('semi') ? semi : perishable)
        : (perishableNames.has(key) ? perishable : semi);
        
      target.push({ name: rawName, percent: pct });
    });

    renderGroups(perishable, semi);
  } catch (err) {
    console.warn('Inventory API failed, using fallback data.', err);
    renderGroups(fallbackData.perishable, fallbackData.semiperishable);
  }
}

// ===== INITIALIZE =====
loadInventory();

File: capstone-full-/inventory.py
from flask import Flask, jsonify, request, render_template
from flask_cors import CORS
import mysql.connector
from datetime import datetime

app = Flask(__name__)
CORS(app)  # Enable CORS for frontend requests

# --- MySQL Connection ---
def get_db_connection():
    return mysql.connector.connect(
        host="localhost",
        user="root",
        password="",
        database="gastrotrack"
    )

# --- Fetch All Inventory (Active & Disabled) ---
@app.route("/api/inventory", methods=["GET"])
def get_inventory():
    """Return all inventory data as JSON with stock levels and status"""
    conn = get_db_connection()
    cursor = conn.cursor(dictionary=True)

    query = """
        SELECT 
            item_id, 
            item_name, 
            stock_level,
            capacity,
            COALESCE(category, 'perishable') AS category,
            COALESCE(status, 'active') AS status,
            unit_cost,
            expiry_date,
            created_at,
            updated_at
        FROM inventory
        ORDER BY status ASC, item_name ASC
    """
    cursor.execute(query)
    items = cursor.fetchall()

    cursor.close()
    conn.close()

    # Compute stock percentage for frontend
    for item in items:
        capacity = float(item.get("capacity", 0))
        stock = float(item.get("stock_level", 0))
        item["stock_percent"] = round((stock / capacity) * 100, 2) if capacity > 0 else 0
        
        # Format dates if present
        if item.get("expiry_date"):
            item["expiry_date"] = item["expiry_date"].strftime("%Y-%m-%d")

    return jsonify(items)


# --- Add New Ingredient ---
@app.route("/api/inventory", methods=["POST"])
def add_ingredient():
    """Add a new ingredient to inventory"""
    data = request.json
    
    if not data.get("item_name"):
        return jsonify({"error": "item_name is required"}), 400
    
    conn = get_db_connection()
    cursor = conn.cursor()

    query = """
        INSERT INTO inventory 
        (item_name, stock_level, capacity, category, unit_cost, status)
        VALUES (%s, %s, %s, %s, %s, 'active')
    """
    
    try:
        cursor.execute(query, (
            data["item_name"],
            data.get("stock_level", 0),
            data.get("capacity", 0),
            data.get("category", "perishable"),
            data.get("unit_cost", 0)
        ))
        conn.commit()
        new_id = cursor.lastrowid
        
        cursor.close()
        conn.close()
        
        return jsonify({
            "message": "Ingredient added successfully",
            "item_id": new_id
        }), 201
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Update Ingredient ---
@app.route("/api/inventory/<int:item_id>", methods=["PUT"])
def update_ingredient(item_id):
    """Update an existing ingredient"""
    data = request.json
    
    conn = get_db_connection()
    cursor = conn.cursor()

    query = """
        UPDATE inventory 
        SET stock_level = %s, 
            capacity = %s, 
            category = %s,
            updated_at = NOW()
        WHERE item_id = %s
    """
    
    try:
        cursor.execute(query, (
            data.get("stock_level"),
            data.get("capacity"),
            data.get("category"),
            item_id
        ))
        conn.commit()
        
        cursor.close()
        conn.close()
        
        return jsonify({"message": "Ingredient updated successfully"})
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Log Restock ---
@app.route("/api/inventory/restock", methods=["POST"])
def log_restock():
    """Log a restock event and update stock level"""
    data = request.json
    
    if not data.get("item_id") or not data.get("quantity"):
        return jsonify({"error": "item_id and quantity are required"}), 400
    
    conn = get_db_connection()
    cursor = conn.cursor()

    # Update stock level by adding the restocked quantity
    query = """
        UPDATE inventory 
        SET stock_level = stock_level + %s,
            updated_at = NOW()
        WHERE item_id = %s
    """
    
    try:
        cursor.execute(query, (data["quantity"], data["item_id"]))
        conn.commit()
        
        cursor.close()
        conn.close()
        
        return jsonify({"message": "Restock logged successfully"})
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Disable Ingredient ---
@app.route("/api/inventory/<int:item_id>/disable", methods=["POST"])
def disable_ingredient(item_id):
    """Disable an ingredient (soft delete)"""
    conn = get_db_connection()
    cursor = conn.cursor()

    query = """
        UPDATE inventory 
        SET status = 'disabled',
            updated_at = NOW()
        WHERE item_id = %s
    """
    
    try:
        cursor.execute(query, (item_id,))
        conn.commit()
        
        cursor.close()
        conn.close()
        
        return jsonify({"message": "Ingredient disabled successfully"})
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Re-enable Ingredient ---
@app.route("/api/inventory/<int:item_id>/enable", methods=["POST"])
def enable_ingredient(item_id):
    """Re-enable a disabled ingredient"""
    conn = get_db_connection()
    cursor = conn.cursor()

    query = """
        UPDATE inventory 
        SET status = 'active',
            updated_at = NOW()
        WHERE item_id = %s
    """
    
    try:
        cursor.execute(query, (item_id,))
        conn.commit()
        
        cursor.close()
        conn.close()
        
        return jsonify({"message": "Ingredient re-enabled successfully"})
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Permanently Delete Ingredient ---
@app.route("/api/inventory/<int:item_id>", methods=["DELETE"])
def delete_ingredient(item_id):
    """Permanently delete an ingredient from database"""
    conn = get_db_connection()
    cursor = conn.cursor()

    query = "DELETE FROM inventory WHERE item_id = %s"
    
    try:
        cursor.execute(query, (item_id,))
        conn.commit()
        
        if cursor.rowcount == 0:
            cursor.close()
            conn.close()
            return jsonify({"error": "Ingredient not found"}), 404
        
        cursor.close()
        conn.close()
        
        return jsonify({"message": "Ingredient permanently deleted"})
        
    except mysql.connector.Error as err:
        cursor.close()
        conn.close()
        return jsonify({"error": str(err)}), 500


# --- Get Low Stock Items ---
@app.route("/api/inventory/low-stock", methods=["GET"])
def get_low_stock():
    """Get items with stock level below 30% capacity"""
    conn = get_db_connection()
    cursor = conn.cursor(dictionary=True)

    query = """
        SELECT 
            item_id, 
            item_name, 
            stock_level,
            capacity,
            category,
            ROUND((stock_level / capacity) * 100, 2) AS stock_percent
        FROM inventory
        WHERE status = 'active'
          AND capacity > 0
          AND (stock_level / capacity) < 0.3
        ORDER BY stock_percent ASC
    """
    
    cursor.execute(query)
    items = cursor.fetchall()

    cursor.close()
    conn.close()

    return jsonify(items)


# --- Frontend Route (Optional) ---
@app.route("/inventory")
def inventory_page():
    """Render inventory management page"""
    return render_template("inventory.html")


# --- Health Check ---
@app.route("/api/health", methods=["GET"])
def health_check():
    """Check if API and database connection are working"""
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute("SELECT 1")
        cursor.close()
        conn.close()
        return jsonify({"status": "healthy", "database": "connected"})
    except Exception as e:
        return jsonify({"status": "unhealthy", "error": str(e)}), 500


if __name__ == "__main__":
    app.run(debug=True, host="127.0.0.1", port=5000)